core/iwasm/compilation/aot_emit_gc.c

```diff
@@ -228,6 +259,1213 @@ aot_compile_op_ref_as_non_null(AOTCompContext *comp_ctx,
+                                   &rtt_type))
+        goto fail;
+
+    ADD_BASIC_BLOCK(check_rtt_type_succ, "check rtt type succ");
+    MOVE_BLOCK_AFTER_CURR(check_rtt_type_succ);
+
+    BUILD_ISNULL(rtt_type, cmp, "cmp_rtt_type");
+    if (!aot_emit_exception(comp_ctx, func_ctx, EXCE_NULL_GC_REF, true, cmp,
+                            check_rtt_type_succ))
+        goto fail;
+
+    /* Generate call wasm_struct_obj_new and check for exception */
+    if (!aot_call_wasm_struct_obj_new(comp_ctx, func_ctx, rtt_type,
+                                      &struct_obj))
+        goto fail;
+
+    ADD_BASIC_BLOCK(check_struct_obj_succ, "check struct obj succ");
+    MOVE_BLOCK_AFTER(check_struct_obj_succ, check_rtt_type_succ);
+
+    BUILD_ISNULL(struct_obj, cmp, "cmp_struct_obj");
+    if (!aot_emit_exception(comp_ctx, func_ctx, EXCE_NULL_GC_REF, true, cmp,
+                            check_struct_obj_succ))
+        goto fail;
+
+    /* For WASM_OP_STRUCT_NEW_CANON, init filed with poped value */
+    if (!init_with_default
+        && !struct_new_canon_init_fields(comp_ctx, func_ctx, type_index,
+                                         struct_obj)) {
+        goto fail;
+    }
+
+    PUSH_REF(struct_obj);
+
+    return true;
+fail:
+    return false;
+}
+
+bool
+aot_compile_op_struct_get(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx,
+                          uint32 type_index, uint32 field_idx, bool sign)
+{
+    LLVMValueRef struct_obj, cmp, field_value;
+    LLVMBasicBlockRef check_struct_obj_succ;
+
+    /* Used in compile time, to distinguish what type of AOTValue PUSH,
+     * field_data offset, size  */
+    WASMStructType *compile_time_struct_type =
+        (WASMStructType *)comp_ctx->comp_data->types[type_index];
+    WASMStructFieldType *field;
+    uint8 field_type, field_size, field_offset;
+
+    field = compile_time_struct_type->fields + field_idx;
+    field_type = field->field_type;
+    field_size = field->field_size;
+    field_offset = field->field_offset;
+
+    if (field_idx >= compile_time_struct_type->field_count) {
+        aot_set_last_error("struct field index out of bounds");
+        goto fail;
+    }
+
+    POP_REF(struct_obj);
+
+    ADD_BASIC_BLOCK(check_struct_obj_succ, "check struct obj succ");
+    MOVE_BLOCK_AFTER_CURR(check_struct_obj_succ);
+
+    BUILD_ISNULL(struct_obj, cmp, "cmp_struct_obj");
+    if (!aot_emit_exception(comp_ctx, func_ctx, EXCE_NULL_GC_REF, true, cmp,
+                            check_struct_obj_succ))
+        goto fail;
+
+    if (!aot_struct_obj_get_field(comp_ctx, struct_obj, I32_CONST(field_offset),
+                                  &field_value, field_size, sign))
+        goto fail;
+
+    if (wasm_is_type_reftype(field_type)) {
+        PUSH_REF(field_value);
+    }
+    else if (field_type == VALUE_TYPE_I32 || field_type == VALUE_TYPE_F32
+             || field_type == PACKED_TYPE_I8 || field_type == PACKED_TYPE_I16) {
+        PUSH_I32(field_value);
```

Comment:
Should zero extend i8/i16 value to i32 if the field_type is i8/i16?

Reply:
The i8/i16 value extended to i32 in `aot_struct_obj_get_field` function: 

```C
if (extend) {
        if (sign_extend) {
            if (!(*field_value = LLVMBuildSExt(comp_ctx->builder, *field_value,
                                               I32_TYPE, "field_value_sext"))) {
                aot_set_last_error("llvm build signed ext failed.");
                goto fail;
            }
        }
        else {
            if (!(*field_value = LLVMBuildZExt(comp_ctx->builder, *field_value,
                                               I32_TYPE, "field_value_zext"))) {
                aot_set_last_error("llvm build unsigned ext failed.");
                goto fail;
            }
        }
    }
```


In interpreter mode, the type conversions happened in `wasm_struct_obj_get_field`. So I did a similar thing.